gosimple and staticcheck linter don't find types defined in other files in the same package.

```

 Current Filetype: go
Available Linters: ['go build', 'gofmt', 'golint', 'gometalinter', 'gosimple', 'go vet', 'staticcheck']
  Enabled Linters: ['gosimple']
 Linter Variables:

 Global Variables:

let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'Error'
let g:ale_echo_msg_format = '%s'
let g:ale_echo_msg_warning_str = 'Warning'
let g:ale_enabled = 1
let g:ale_fix_on_save = 0
let g:ale_fixers = {}
let g:ale_keep_list_window_open = 0
let g:ale_lint_delay = 200
let g:ale_lint_on_enter = 1
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 'always'
let g:ale_linter_aliases = {}
let g:ale_linters = {'go': ['gosimple']}
let g:ale_open_list = 0
let g:ale_set_highlights = 1
let g:ale_set_loclist = 1
let g:ale_set_quickfix = 0
let g:ale_set_signs = 1
let g:ale_sign_column_always = 0
let g:ale_sign_error = '>>'
let g:ale_sign_offset = 1000000
let g:ale_sign_warning = '⚠'
let g:ale_statusline_format = ['%d error(s)', '%d warning(s)', 'OK']
let g:ale_warn_about_trailing_whitespace = 1
  Command History:

(executable check - success) gosimple
(started) ['bash', '-c', 'gosimple ''/tmp/nvimuKWnOb/4/raft.go''']
(started) ['bash', '-c', 'gosimple ''/tmp/nvimuKWnOb/7/raft.go''']
(started) ['bash', '-c', 'gosimple ''/tmp/nvimuKWnOb/8/raft.go''']
(started) ['bash', '-c', 'gosimple ''/tmp/nvimuKWnOb/10/raft.go''']
(finished - exit code 1) ['bash', '-c', 'gosimple ''/tmp/nvimuKWnOb/11/raft.go''']

<<<OUTPUT STARTS>>>
/tmp/nvimuKWnOb/11/raft.go:25:11: undeclared name: remoteKVDatabase
/tmp/nvimuKWnOb/11/raft.go:33:14: undeclared name: remoteKVConnection
/tmp/nvimuKWnOb/11/raft.go:125:4: undeclared name: KVDatabase
/tmp/nvimuKWnOb/11/raft.go:150:65: undeclared name: remoteKVConnection
/tmp/nvimuKWnOb/11/raft.go:179:51: undeclared name: remoteKVConnection
/tmp/nvimuKWnOb/11/raft.go:192:49: undeclared name: remoteKVConnection
/tmp/nvimuKWnOb/11/raft.go:279:56: undeclared name: KVTable
/tmp/nvimuKWnOb/11/raft.go:289:42: undeclared name: Row
/tmp/nvimuKWnOb/11/raft.go:313:42: undeclared name: SeekAt
/tmp/nvimuKWnOb/11/raft.go:313:99: undeclared name: Row
/tmp/nvimuKWnOb/11/raft.go:326:51: undeclared name: SelectArgs
/tmp/nvimuKWnOb/11/raft.go:326:89: undeclared name: Row
/tmp/nvimuKWnOb/11/raft.go:43:72: undeclared name: KVManager
/tmp/nvimuKWnOb/11/raft.go:117:30: invalid operation: m.conn (variable of type *invalid type) has no field or method conn
/tmp/nvimuKWnOb/11/raft.go:128:18: undeclared name: newRemoteKVDatabase
/tmp/nvimuKWnOb/11/raft.go:155:15: undeclared name: newRemoteKVConnection
/tmp/nvimuKWnOb/11/raft.go:195:12: undeclared name: remoteKVConnection
/tmp/nvimuKWnOb/11/raft.go:280:9: undeclared name: newKVTable
/tmp/nvimuKWnOb/11/raft.go:285:10: invalid operation: db.remote (variable of type *invalid type) has no field or method Put
/tmp/nvimuKWnOb/11/raft.go:291:10: invalid operation: db.remote (variable of type *invalid type) has no field or method PutMany
/tmp/nvimuKWnOb/11/raft.go:300:14: invalid operation: db.remote (variable of type *invalid type) has no field or method Get
/tmp/nvimuKWnOb/11/raft.go:309:10: invalid operation: db.remote (variable of type *invalid type) has no field or method Delete
/tmp/nvimuKWnOb/11/raft.go:315:13: undeclared name: Row
/tmp/nvimuKWnOb/11/raft.go:319:15: invalid operation: db.remote (variable of type *invalid type) has no field or method Select
/tmp/nvimuKWnOb/11/raft.go:328:20: undeclared name: Row
/tmp/nvimuKWnOb/11/raft.go:332:20: invalid operation: db.remote (variable of type *invalid type) has no field or method BatchSelect
/tmp/nvimuKWnOb/11/raft.go:340:2: invalid operation: db.remote (variable of type *invalid type) has no field or method Close
/tmp/nvimuKWnOb/11/raft.go:345:10: invalid operation: db.remote (variable of type *invalid type) has no field or method FlushAndClose
/tmp/nvimuKWnOb/11/raft.go:353:14: invalid operation: db.remote (variable of type *invalid type) has no field or method GetProperty
/tmp/nvimuKWnOb/11/raft.go:60:34: invalid operation: m.conn (variable of type *invalid type) has no field or method conn
couldn't load packages due to errors: adhoc
<<<OUTPUT ENDS>>>

```

I get the same output when I only run the `gosimple` command on the single file I'm in. Running it on the full directory works fine though. Is there a way to define the linter in such a way that the temporary directory with the new file contains the other files in the directory as well?

--- a/ale_linters/go/gosimple.vim
+++ b/ale_linters/go/gosimple.vim
@@ -9,7 +9,7 @@ call ale#linter#Define('go', {
 \   'name': 'gosimple',
 \   'executable': 'gosimple',
 \   'command_callback': 'ale_linters#go#gosimple#GetCommand',
-\   'callback': 'ale#handlers#unix#HandleAsWarning',
+\   'callback': 'ale#handlers#go#Handler',
 \   'output_stream': 'both',
 \   'lint_file': 1,
 \})

--- a/ale_linters/go/gosimple.vim
+++ b/ale_linters/go/gosimple.vim
@@ -1,10 +1,14 @@
 " Author: Ben Reedy <https://github.com/breed808>
 " Description: gosimple for Go files
 
+function! ale_linters#go#gosimple#GetCommand(buffer) abort
+    return ale#path#BufferCdString(a:buffer) . ' gosimple .'
+endfunction
+
 call ale#linter#Define('go', {
 \   'name': 'gosimple',
 \   'executable': 'gosimple',
-\   'command': 'gosimple %s',
+\   'command_callback': 'ale_linters#go#gosimple#GetCommand',
 \   'callback': 'ale#handlers#unix#HandleAsWarning',
 \   'output_stream': 'both',
 \   'lint_file': 1,

--- a/ale_linters/go/gotype.vim
+++ b/ale_linters/go/gotype.vim
@@ -15,6 +15,6 @@ call ale#linter#Define('go', {
 \   'output_stream': 'stderr',
 \   'executable': 'gotype',
 \   'command_callback': 'ale_linters#go#gotype#GetCommand',
-\   'callback': 'ale#handlers#unix#HandleAsError',
+\   'callback': 'ale#handlers#go#Handler',
 \   'lint_file': 1,
 \})

--- a/ale_linters/go/gotype.vim
+++ b/ale_linters/go/gotype.vim
@@ -6,7 +6,8 @@ function! ale_linters#go#gotype#GetCommand(buffer) abort
         return
     endif
 
-    return 'gotype %s'
+
+    return ale#path#BufferCdString(a:buffer) . ' gotype .'
 endfunction
 
 call ale#linter#Define('go', {

--- a/ale_linters/go/govet.vim
+++ b/ale_linters/go/govet.vim
@@ -8,28 +8,11 @@ function! ale_linters#go#govet#GetCommand(buffer) abort
     return ale#path#BufferCdString(a:buffer) . ' go vet .'
 endfunction
 
-function! ale_linters#go#govet#Handler(buffer, lines) abort
-    let l:pattern = '\v^([a-zA-Z]?:?[^:]+):(\d+):?(\d+)?:? ?(.+)$'
-    let l:output = []
-    let l:dir = expand('#' . a:buffer . ':p:h')
-
-    for l:match in ale#util#GetMatches(a:lines, l:pattern)
-        call add(l:output, {
-        \   'filename': ale#path#GetAbsPath(l:dir, l:match[1]),
-        \   'lnum': l:match[2] + 0,
-        \   'col': l:match[3] + 0,
-        \   'text': l:match[4],
-        \   'type': 'E',
-        \})
-    endfor
-    return l:output
-endfunction
-
 call ale#linter#Define('go', {
 \   'name': 'go vet',
 \   'output_stream': 'stderr',
 \   'executable': 'go',
 \   'command_callback': 'ale_linters#go#govet#GetCommand',
-\   'callback': 'ale_linters#go#govet#Handler',
+\   'callback': 'ale#handlers#go#Handler',
 \   'lint_file': 1,
 \})

--- a/ale_linters/go/staticcheck.vim
+++ b/ale_linters/go/staticcheck.vim
@@ -27,7 +27,7 @@ call ale#linter#Define('go', {
 \   'name': 'staticcheck',
 \   'executable': 'staticcheck',
 \   'command_callback': 'ale_linters#go#staticcheck#GetCommand',
-\   'callback': 'ale#handlers#unix#HandleAsWarning',
+\   'callback': 'ale#handlers#go#Handler',
 \   'output_stream': 'both',
 \   'lint_file': 1,
 \})

--- a/autoload/ale/handlers/go.vim
+++ b/autoload/ale/handlers/go.vim
@@ -1,0 +1,25 @@
+" Author: neersighted <bjorn@neersighted.com>
+" Description: go vet for Go files
+"
+" Author: John Eikenberry <jae@zhar.net>
+" Description: updated to work with go1.10
+"
+" Author: Ben Paxton <ben@gn32.uk>
+" Description: moved to generic Golang file from govet
+
+function! ale#handlers#go#Handler(buffer, lines) abort
+    let l:pattern = '\v^([a-zA-Z]?:?[^:]+):(\d+):?(\d+)?:? ?(.+)$'
+    let l:output = []
+    let l:dir = expand('#' . a:buffer . ':p:h')
+
+    for l:match in ale#util#GetMatches(a:lines, l:pattern)
+        call add(l:output, {
+        \   'filename': ale#path#GetAbsPath(l:dir, l:match[1]),
+        \   'lnum': l:match[2] + 0,
+        \   'col': l:match[3] + 0,
+        \   'text': l:match[4],
+        \   'type': 'E',
+        \})
+    endfor
+    return l:output
+endfunction

--- a/test/command_callback/test_gosimple_command_callback.vader
+++ b/test/command_callback/test_gosimple_command_callback.vader
@@ -1,0 +1,12 @@
+Before:
+  runtime ale_linters/go/gosimple.vim
+  call ale#test#SetFilename('../go_files/testfile2.go')
+
+After:
+  call ale#linter#Reset()
+
+Execute(The default gosimple command should be correct):
+  AssertEqual 'cd ' . ale#Escape(expand('%:p:h')) . ' && '
+  \ . ' gosimple .',
+  \ ale_linters#go#gosimple#GetCommand(bufnr(''))
+

--- a/test/command_callback/test_gotype_command_callback.vader
+++ b/test/command_callback/test_gotype_command_callback.vader
@@ -6,7 +6,9 @@ After:
   call ale#linter#Reset()
 
 Execute(The default gotype command should be correct):
-  AssertEqual 'gotype %s', ale_linters#go#gotype#GetCommand(bufnr(''))
+  AssertEqual 'cd ' . ale#Escape(expand('%:p:h')) . ' && '
+  \ . ' gotype .',
+  \ ale_linters#go#gotype#GetCommand(bufnr(''))
 
 Execute(The gotype callback should ignore test files):
   call ale#test#SetFilename('bla_test.go')

--- a/test/handler/test_go_generic_handler.vader
+++ b/test/handler/test_go_generic_handler.vader
@@ -1,10 +1,4 @@
-Before:
-  runtime ale_linters/go/govet.vim
-
-After:
-  call ale#linter#Reset()
-
-Execute(The govet handler should return the correct filenames):
+Execute(The golang handler should return the correct filenames):
   AssertEqual
   \ [
   \   {
@@ -22,7 +16,7 @@ Execute(The govet handler should return the correct filenames):
   \     'filename': ale#path#Simplify(expand('%:p:h') . '/other.go'),
   \   },
   \ ],
-  \ ale_linters#go#govet#Handler(bufnr(''), [
+  \ ale#handlers#go#Handler(bufnr(''), [
   \   'test.go:27: some error',
   \   'other.go:27:5: some error with a column',
   \ ])

